src/hash_value.cpp

```diff
@@ -371,6 +371,56 @@ bool HashValue::eql(Env *env, ValuePtr other_value) {
     return eq(env, other_value, SymbolValue::intern("eql?"));
 }
 
+bool HashValue::gte(Env *env, ValuePtr other) {
+    if (!other->is_hash() && other->respond_to_method(env, SymbolValue::intern("to_hash")))
+        other = other->send(env, SymbolValue::intern("to_hash"));
+
+    other->assert_type(env, Value::Type::Hash, "Hash");
+    auto other_hash = other->as_hash();
+
+    for (auto &node : *other_hash) {
+        ValuePtr value = get(env, node.key);
+        if (!value || value.send(env, SymbolValue::intern("=="), { node.val })->is_false()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool HashValue::gt(Env *env, ValuePtr other) {
+    if (!other->is_hash() && other->respond_to_method(env, SymbolValue::intern("to_hash")))
+        other = other->send(env, SymbolValue::intern("to_hash"));
+
+    other->assert_type(env, Value::Type::Hash, "Hash");
+
+    return gte(env, other) && other->as_hash()->size() != size();
+}
+
+bool HashValue::lte(Env *env, ValuePtr other) {
+    if (!other->is_hash() && other->respond_to_method(env, SymbolValue::intern("to_hash")))
+        other = other->send(env, SymbolValue::intern("to_hash"));
+
+    other->assert_type(env, Value::Type::Hash, "Hash");
+    auto other_hash = other->as_hash();
+
+    for (auto &node : *this) {
+        ValuePtr value = other_hash->get(env, node.key);
+        if (!value || value.send(env, SymbolValue::intern("=="), { node.val })->is_false()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool HashValue::lt(Env *env, ValuePtr other) {
+    if (!other->is_hash() && other->respond_to_method(env, SymbolValue::intern("to_hash")))
+        other = other->send(env, SymbolValue::intern("to_hash"));
+
+    other->assert_type(env, Value::Type::Hash, "Hash");
+
+    return lte(env, other) && other->as_hash()->size() != size();
```

Comment:
Can you explain how this works? At first glance, it seems wrong to me...

One of the specs says `it 'returns true when receiver is a subset of argument'` and this doesn't seem to check for that?

Reply:
One hash a is "less than" another hash b if each element of a is in b (checked by `lte`) and they are not equal. The trick is instead of checking for equality we just test if the size of a and b are different because if they are and all elements of a are in b (which we checked by `lte`), they must be equal. Does this make any sense? 😅 

Do you have an idea of making the code more understandable?

Reply:
No I think I get it. Works for me!